### PR TITLE
BUG: DataFrame(dict_of_series) raising depending on order of dict

### LINKED ISSUE #44091: BUG: DataFrame(dict_of_series) raises depending on order of dict
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
dti = pd.date_range("2016-01-01", periods=3)
ser1 = pd.Series(range(3), index=dti)
ser2 = pd.Series(range(3), index=dti.tz_localize("UTC"))
ser3 = pd.Series(range(3))

>>> pd.DataFrame({"A": ser3, "B": ser2, "C": ser1})  # works fine

>>> pd.DataFrame({"A": ser1, "B": ser2, "C": ser3})
TypeError: Cannot join tz-naive with tz-aware DatetimeIndex
```

### Issue Description

When we pass a dict of Series, we end up going through indexes.api.union_indexes, which checks if our _first_ index is a DatetimeIndex and if so dispatches to DatetimeIndex.union_many, which in this case raises.  If the DatetimeIndex is not first, then it goes through Index.union and casts to object.

Then there's the issue that we don't pass `sort` to union_many so it gets ignored.

### Expected Behavior

Cast to object.

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #44091
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/api.py [hunk 1]
```diff
@@ -213,14 +215,41 @@ def conv(i):
 
     if kind == "special":
         result = indexes[0]
+        first = result
```
Comment (jreback) [inline pandas/core/indexes/api.py:218]: can you move all of this stuff out of api.py (e.g. new module), can be followon

File: pandas/core/indexes/api.py [hunk 2]
```diff
@@ -213,14 +215,41 @@ def conv(i):
...
+
+        dtis = [x for x in indexes if isinstance(x, DatetimeIndex)]
```
Comment (jreback) [inline pandas/core/indexes/api.py:220]: this is super special casey. umm

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/api.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/api.py:220]: yah its kinda gross but the best i could figure out.  it gets a bit better once a deprecation is enforced in 2.0, and we could consider changing the raising case

File: pandas/core/indexes/api.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/api.py:218]: will give some thought to a proper home

File: pandas/core/indexes/api.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/api.py:220]: ok will look in a bit more detail

File: pandas/core/indexes/api.py [hunk 3]
```diff
@@ -213,14 +215,41 @@ def conv(i):
...
+        dti_tzs = [x for x in dtis if x.tz is not None]
+        if len(dti_tzs) not in [0, len(dtis)]:
```
Comment (jreback) [inline pandas/core/indexes/api.py:222]: can we not just kill union_many now (and just have it call union_indexes)?

Comment (jbrockmendel) [reply pandas/core/indexes/api.py:222]: that would change the behavior of union_many, which in principle would require a deprecation

Comment (jreback) [reply pandas/core/indexes/api.py:222]: ok, but we can just deprecate this anyhow (union_many)? as you have seemed to replaced it

Comment (jreback) [reply pandas/core/indexes/api.py:222]: to be honest i am ok with just killing it and replacing the code path. its annoying keeping these 2 different paths.

Comment (jbrockmendel) [reply pandas/core/indexes/api.py:222]: This _is_ deprecating union_many

Comment (jreback) [reply pandas/core/indexes/api.py:222]: ok yeah i guess deprecation is ok

**Comment by jbrockmendel** [comment]
greenish

**Review by jreback — APPROVED** [review]

**Pull request merged**